staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go

```diff
@@ -484,3 +490,192 @@ func TestCacheWatcherStoppedInAnotherGoroutine(t *testing.T) {
+				if expectedBookmarks {
+					rv, err := cacher.versioner.ObjectResourceVersion(event.Object)
+					if err != nil {
+						t.Errorf("failed to parse resource version from %#v", event.Object)
+					}
+					if rv != resourceVersion {
+						t.Errorf("expected a bookmark event with a resource version %v (actual %v)", resourceVersion, rv)
+					}
+					return
+				}
+				t.Fatalf("expected never received bookmark events")
+			}
+		case <-time.After(2 * time.Second):
+			if expectedBookmarks {
+				t.Fatalf("expected received bookmark events (err: timeout)")
+			}
+			return
+		}
+	}
+}
+
+func TestCacherSendBookmarkEvents(t *testing.T) {
+	testCases := []struct {
+		watchCacheEnabled   bool
+		allowWatchBookmarks bool
+		expectedBookmarks   bool
+	}{
+		{
+			watchCacheEnabled:   true,
+			allowWatchBookmarks: true,
+			expectedBookmarks:   true,
+		},
+		{
+			watchCacheEnabled:   true,
+			allowWatchBookmarks: false,
+			expectedBookmarks:   false,
+		},
+		{
+			watchCacheEnabled:   false,
+			allowWatchBookmarks: true,
+			expectedBookmarks:   false,
+		},
+		{
+			watchCacheEnabled:   false,
+			allowWatchBookmarks: false,
+			expectedBookmarks:   false,
+		},
+	}
+
+	for _, tc := range testCases {
+		testCacherSendBookmarkEvents(t, tc.watchCacheEnabled, tc.allowWatchBookmarks, tc.expectedBookmarks)
+	}
+}
+
+func TestDispatchingBookmarkEventsWithConcurrentStop(t *testing.T) {
+	defer utilfeaturetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.WatchBookmark, true)()
+	backingStorage := &dummyStorage{}
+	cacher, _ := newTestCacher(backingStorage, 1000)
+	defer cacher.Stop()
+
+	// Wait until cacher is initialized.
+	cacher.ready.wait()
+
+	// Ensure there is some budget for slowing down processing.
+	cacher.dispatchTimeoutBudget.returnUnused(100 * time.Millisecond)
+
+	resourceVersion := uint64(1000)
+	err := cacher.watchCache.Add(&examplev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            fmt.Sprintf("pod-0"),
+			Namespace:       "ns",
+			ResourceVersion: fmt.Sprintf("%v", resourceVersion),
+		}})
+	if err != nil {
+		t.Fatalf("failed to add a pod: %v", err)
+	}
+
+	for i := 0; i < 1000; i++ {
+		w, err := cacher.Watch(context.TODO(), "pods/ns", "999", storage.Everything)
+		if err != nil {
+			t.Fatalf("Failed to create watch: %v", err)
+		}
+		bookmark := &watchCacheEvent{
+			Type:            watch.Bookmark,
+			ResourceVersion: uint64(i),
+			Object:          cacher.newFunc(),
+		}
+		cacher.versioner.UpdateObject(bookmark.Object, bookmark.ResourceVersion)
+
+		go func() {
+			cacher.dispatchEvent(bookmark)
+		}()
+
+		go func() {
+			w.Stop()
+		}()
+
+		receiveResult := func() <-chan struct{} {
+			done := make(chan struct{})
+			go func() {
+				for range w.ResultChan() {
+				}
+				close(done)
+			}()
+			return done
+		}
+
+		select {
+		case <-receiveResult():
+			break
+		case <-time.After(time.Second):
+			t.Errorf("receive result timeout")
+		}
+		w.Stop()
+	}
+}
```

Comment:
One test that I'm missing here is verifying whether bookmarks have correct RVs.
The test may be something like:
- generate a stream of pods with consecutive RVs
- open a watch with say 1s deadline
- verify that when it received bookmark its RV was not lower than RV of previous object, and next object will have higher RV

Reply:
done

Reply:
awesome - thanks!